chain/network/src/raw.rs

```diff
@@ -0,0 +1,345 @@
+            if first_byte_time.is_none() {
+                first_byte_time = Some(Instant::now());
+            }
+        }
+        let len = u32::from_le_bytes(self.buf[..4].try_into().unwrap());
+        // If first_byte_time is None, there were already 4 bytes from last time,
+        // and they must have come before a partial frame.
+        // So the Instant::now() is not quite correct, since the time was really in the past,
+        // but this is prob not so important
+        Ok((len as usize, first_byte_time.unwrap_or(Instant::now())))
+    }
+
+    // drains anything still available to be read without blocking and
+    // returns a vec of message lengths currently ready
+    fn read_remaining_messages(
+        &mut self,
+        first_msg_length: usize,
+    ) -> io::Result<Vec<(usize, Instant)>> {
+        let mut pos = first_msg_length + 4;
+        let mut lengths = Vec::new();
+        let mut pending_msg_length = None;
+
+        loop {
+            match self.stream.try_read_buf(&mut self.buf) {
+                Ok(n) => {
+                    if n == 0 {
+                        return Err(io::Error::new(
+                            io::ErrorKind::UnexpectedEof,
+                            "no more bytes available, but expected to receive a message",
+                        ));
+                    }
+                    let timestamp = Instant::now();
+                    tracing::trace!(target: "network", "Read {} bytes from {:?} non-blocking", n, self.stream.peer_addr());
+                    loop {
+                        if pending_msg_length.is_none() && self.buf.remaining() - pos >= 4 {
+                            let len =
+                                u32::from_le_bytes(self.buf[pos..pos + 4].try_into().unwrap());
+                            pending_msg_length = Some(len as usize);
+                        }
+                        if let Some(l) = pending_msg_length {
+                            if self.buf.remaining() - pos >= l + 4 {
+                                lengths.push((l, timestamp));
+                                pos += l + 4;
+                                pending_msg_length = None;
+                            } else {
+                                break;
+                            }
+                        } else {
+                            break;
+                        }
+                    }
+                }
+                Err(e) => match e.kind() {
+                    io::ErrorKind::WouldBlock => {
+                        break;
+                    }
+                    _ => return Err(e),
+                },
+            };
+        }
+        Ok(lengths)
+    }
+
+    // Reads from the socket until there is at least one full PeerMessage available.
+    // After that point, continues to read any bytes available to be read without blocking
+    // and appends any extra messages to the returned Vec. Usually there is only one in there
```

Comment:
what's the purpose of this semantics?

Reply:
Ok so I did this originally because if you look at the trace level logs, you'll see that we do read more messages without blocking a lot of the time, and what I wanted to avoid was a situation where we get a big message, take a long time to deserialize it (in the case of a big `SyncRoutingTable`, 100ms+), and then report distorted latencies for subsequent pong messages. But looking at the logs some more, it's pretty clear that that only happens much at the beginning, and after that, we're talking about differences in reported latencies of < 1ms, so sure, I agree, let's just keep it simpler